core/corehttp/routing.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/ipfs/boxo/routing/http/types/iter"
+	cid "github.com/ipfs/go-cid"
+	core "github.com/ipfs/kubo/core"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+)
+
+func RoutingOption() ServeOption {
+	return func(n *core.IpfsNode, _ net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
+		handler := server.Handler(&contentRouter{n})
+		mux.Handle("/routing/v1/", handler)
+		return mux, nil
+	}
+}
+
+type contentRouter struct {
+	n *core.IpfsNode
+}
+
+func (r *contentRouter) FindProviders(ctx context.Context, key cid.Cid, limit int) (iter.ResultIter[types.Record], error) {
+	ctx, cancel := context.WithCancel(ctx)
+	ch := r.n.Routing.FindProvidersAsync(ctx, key, limit)
+	return iter.ToResultIter[types.Record](&peerChanIter{
+		ch:     ch,
+		cancel: cancel,
+	}), nil
+}
+
+// nolint deprecated
+func (r *contentRouter) ProvideBitswap(ctx context.Context, req *server.BitswapWriteProvideRequest) (time.Duration, error) {
+	return 0, routing.ErrNotSupported
+}
+
+func (r *contentRouter) FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[types.Record], error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	addr, err := r.n.Routing.FindPeer(ctx, pid)
+	if err != nil {
+		return nil, err
+	}
+
+	rec := &types.PeerRecord{
+		Schema: types.SchemaPeer,
+		ID:     &addr.ID,
+	}
+
+	for _, addr := range addr.Addrs {
+		rec.Addrs = append(rec.Addrs, types.Multiaddr{Multiaddr: addr})
+	}
+
+	return iter.ToResultIter[types.Record](iter.FromSlice[types.Record]([]types.Record{rec})), nil
+}
+
+func (r *contentRouter) FindIPNS(ctx context.Context, name ipns.Name) (*ipns.Record, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	raw, err := r.n.Routing.GetValue(ctx, string(name.RoutingKey()))
+	if err != nil {
+		return nil, err
+	}
+
+	return ipns.UnmarshalRecord(raw)
+}
+
+func (r *contentRouter) ProvideIPNS(ctx context.Context, name ipns.Name, record *ipns.Record) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	raw, err := ipns.MarshalRecord(record)
+	if err != nil {
+		return err
+	}
+
+	// The caller guarantees that name matches the record. This is double checked
+	// by the internals of PutValue.
+	return r.n.Routing.PutValue(ctx, string(name.RoutingKey()), raw)
+}
+
+type peerChanIter struct {
+	ch     <-chan peer.AddrInfo
+	cancel context.CancelFunc
+	next   *peer.AddrInfo
+}
+
+func (it *peerChanIter) Next() bool {
+	addr, ok := <-it.ch
+	if ok {
+		it.next = &addr
+		return true
+	} else {
+		it.next = nil
+		return false
+	}
+}
+
+func (it *peerChanIter) Val() types.Record {
+	if it.next == nil {
+		return nil
+	}
+
+	rec := &types.PeerRecord{
+		Schema: types.SchemaPeer,
+		ID:     &it.next.ID,
+		// We need to add this here, otherwise contentrouter will ignore the peer
+		// as it does not have transport bitswap available.
+		Protocols: []string{"transport-bitswap"},
```

Comment:
@lidel This is important to note. Relates to https://github.com/ipfs/boxo/pull/422#discussion_r1305340163.

Reply:
Good question. Soon-ish, Bitswap will no longer be the only thing available:
- In near future we will attempt to use HTTP transport (trustless gateway) when `/http*` multiaddr is present, and such provider result may not support bitswap at all (https://github.com/ipni/indexstar/issues/147)
- We also see slow switch away from graphsync, which is always returned with own schema. These provider results will probably switch to HTTP at some point, Filecoin SPs already started exposing it (https://boost.filecoin.io/http-retrieval).
- There are valid scenarios where there is no `Protocols` and only `ID` is returned. 
  - We want to stop faking bitswap on DHT proxied results when we don't actually know (that requires connect and identify).
  - Peer may listen on multiple protocols, and simply not announce everything to IPNI, or be a result proxied from DHT.
  - That is still useful for Kubo, as it might do additional peer lookup on DHT to learn more about peer, or even connect to it if there is no better option. 

@hacdias  I'd say we should stop filtering out of anything other than bitswap when it comes to the `peer` schema, 
remove things like https://github.com/ipfs/boxo/blob/4871056dde8606d899102d4935d66b2f854960bc/routing/http/contentrouter/contentrouter.go#L138-L141

Reply:
Done!